aten/src/ATen/native/cpu/Parallel.h

```diff
@@ -7,18 +8,33 @@
 namespace at {
 namespace native {
 
-constexpr size_t _THRESHOLD = 32768;
+// This parameter is heuristically chosen to determine the minimum number of
+// work that warrants paralellism. For example, when summing an array, it is
+// deemed inefficient to parallelise over arrays shorter than 32768. Further,
+// no parallel algorithm (such as parallel_reduce) should split work into
+// smaller than GRAIN_SIZE chunks.
+constexpr size_t GRAIN_SIZE = 32768;
 
 template <class T, template <class> class PRED>
 T parallel_reduce(T (*f)(const T *, size_t, size_t, T), const T *data,
                   size_t start, size_t end, T init_) {
+
+  static tbb::task_scheduler_init tbbinit;
```

Comment:
This doesn't look right. Each instantiation of `parallel_reduce` will have its own copy of `tbb::task_scheduler_init tbbinit`.

I still don't fully understand how `tbb::task_scheduler_init` works, but I don't think these will stack properly.

Reply:
Also, it looks like `task_scheduler_init` is mostly meant to initialize some thread-local variables, so making it `static` might not be a good idea.

Reply:
What could be a better approach?

Reply:
Using thread_local static variables seem to do the trick. The thread_local keyword makes sure that there will be a new variable per thread and also removes any need to guard for data races. The static variable makes sure it's linked internally. 

On task_scheduler_init, it needs to be called before the first usage of any tbb function. It initializes the scheduler with the passed parameters (number of threads). Any tbb construct (such as the parallel algorithm) will spawn a default thread pool if this hasn't been called before. The default will consume all available cores. See [the documentation](https://www.threadingbuildingblocks.org/docs/help/reference/task_scheduler/task_scheduler_init_cls.html):

"To override the automatic defaults for task scheduling, a task_scheduler_init must become active before the first use of task scheduling services.
"
There are two potential issues:
We must direct any usage of tbb to Parallel.h to make sure tbb is initialized first.
While the current scheduler might be limited in the number of cores used, the overall market is not. That means, if there are two threads, they each will spawn the number of threads as specified by torch.set_num_thread. As far as I know this behavior is consistent with OpenMP.

One advantage, as far as I can tell, is that tbb will use a market to limited the number of overall worker threads. So that even if each cpp thread requests 60 threads for each tbb scheduler, the overall number of workers will be limited to the maximum number of cores available and the associated number of worker threads is divided by the number of active schedulers. This should prevent oversubscription. 

See https://software.intel.com/en-us/blogs/2011/04/09/tbb-initialization-termination-and-resource-management-details-juicy-and-gory